vere: chunk event log into epochs

mcevoypeter opened the pull request:
Split the event log into epochs, which will enable event log truncation. 

The new directory hierarchy of the `<pier>/.urb/log` directory looks like:
```console
dev/.urb/log
├── 0i1
│   ├── data.mdb
│   ├── lifecycle.bin
│   ├── lock.mdb
│   └── version.bin
├── 0i112
│   ├── data.mdb
│   ├── lock.mdb
│   ├── north.bin
│   ├── south.bin
│   └── version.bin
├── 0i201
│   ├── data.mdb
│   ├── lock.mdb
│   ├── north.bin
│   ├── south.bin
│   └── version.bin
├── fake.bin
└── who.bin
```
Each epoch contains a `version.bin` file specifying the epoch version it represents as well as an LMDB instance for storing the epoch's events. The first epoch `0i1`, has `lifecycle.bin`, which specifies the length of the boot sequence. All non-first epochs have snapshots representing the state of the system immediately before the epoch's first event was applied (i.e. applying the snapshot in `0i112` is equivalent to replaying up to and including event `111`).

Migration from the old event log format is in place so long as the incremental snapshot is up to date with the event log. If not, instructions for how to prepare for migration are printed to the console should migration fail. Rollover to a new epoch occurs immediately after migration.

Rollover to a new epoch is dictated by a fixed max event count for an epoch; when the max event count is reached, a new epoch is automatically created. 

Key additions:
- `epoc` module: manages the persistence of an epoch behind a simple interface. The storage method is still LMDB, but nothing about the `epoc` interface leaks that information, and so it should be fairly easy to swap in a different database should the need arise.
- `evlo` module: manages the epochs that comprise the event log behind a simple interface and provides both synchronous and asynchronous commit options. Similar to the `epoc` interface, the `evlo` interface hides the existence of epochs from the users, which enables greater flexibility should the need to change the event log structure arise sometime in the future.

Key dependencies:
- `list` module: doubly-linked list abstraction.
- `path` module: file path abstraction.
- `bile` module: binary file abstraction.

Remaining loose ends:
- Address the nuances at the intersection of truncation and replaying the boot sequence. As things currently stand, the boot sequence must be replayed if the incremental snapshot is absent, which is clearly problematic if the first epoch, which contains the boot sequence has been dropped.
- Decide on an appropriate max event count for epochs.

Next steps:
- Decide on a truncation policy.

Testing:
- Migration:
```console
$ urbit-1.8 -F dev
$ urbit-peter-epoch dev
```
- Replay:
```console
$ rm dev/.urb/chk/*.bin
$ urbit-peter-epoch dev
```
- Rollover: occurs automatically after a fixed number of events. Checked with `ls dev/.urb/log`. Both replay and migration were tested when a single epoch (i.e. `0i1`) was present and when multiple epochs (up to ten of them) were present.

joemfb commented:
This is great! I'll come back with a more detailed review soon, but wanted to get these notes to you ASAP.

This PR removes some functionality (some of which was added/restored while your branch was in progress):

- lockfile management: in `vere/disk.c` (moved from `vere/io/unix.c`), now additionally used in `worker/main.c`. this is likely causing the CI issues
- commit batch size histogram: in `vere/disk.c`. it's fine to disable the printf on exit (I wasn't going to ship it), but we really want to capture this info,and at least be able to print it from a debugger. Eventually, it'll be wired into the `%info` system, accessible by a signal, &c.
- partial replay: I only just reimplemented this on `jb/wip-urth`. It's not much code, and should be straightforward to port here.

We'll need to refactor some of these `mars.c` changes. Separate functions for `boot` and `replay` are important, and we don't want to be using implementation-detail macros like `try_epoch` across module boundaries. Might be easiest for me to make a PR on top of yours for this sort of thing, but I wanted to mention it now.

Finally, the new module names also aren't going to fly. While we occasionally drift (`lmdb`, `http`), it's important to aim for meaningful, 4-letter english words. Some ideas:

- disk, rock (hard disk, stability, persistence)
- tale, saga, epic (story, history. grand heroic narrative, connects to epoch)
- anon (following one after another)
- fact, feat, know

joemfb commented on pkg/urbit/vere/lock.c:
```diff
@@ -0,0 +1,80 @@
+  }
+
+  fprintf(lok_u, "%u\n", getpid());
+  c3_sync(fileno(lok_u));
```
The macos build is failing because this needs `fcntl.h`

mcevoypeter replied:
Should `#include <fcntl.h>` be included directly here or in `portable.h` within the appropriate `#if defined(U3_OS_osx)` block?

joemfb commented on pkg/urbit/vere/db/lmdb.c:
```diff
@@ -1,627 +0,0 @@
-  fstat(han_u, &sat_u);
-  return (intmax_t)sat_u.st_size;
-}
-#endif
```
The mingw build is failing because of issues with `mdb_strerror()` -- you'll need to replicate this pattern with `mdb_logerror()`

joemfb replied:
For context, see `compat/mingw/lmdb.patch`.

joemfb reviewed (changes requested):
More info about CI failures:

mcevoypeter commented:
A summary of the latest batch of commits that I pushed:
1. Address CI failures.
2. Change event log module name from `evlo` to `saga`.
3. Add support for partial replay.
4. Add support for truncation by only running the boot sequence when the first epoch is the only epoch present.

mcevoypeter reviewed (commented):
Bookmark: review of `c3` changes complete.

mcevoypeter commented on Makefile:
```diff
@@ -1,10 +1,10 @@
 
 build:
-	nix-build -A urbit -A herb --no-out-link
+	nix-build -A urbit --no-out-link
 
 install:
-	nix-env -f . -iA urbit -iA herb
+	nix-env -f . -iA urbit
```
Done.

mcevoypeter commented on pkg/urbit/c/bile.c:
```diff
@@ -0,0 +1,221 @@
+  ssize_t rea_i;
+  while ( 0 < lef_i ) {
+    rea_i = read(bil_u->fid_i, dat_y, lef_i);
+    if ( -1 == rea_i ) {
+      if ( EINTR == errno ) { // (2)
+        continue;
+      }
+      else {
+        goto abandon_read;
+      }
+    }
+    else if ( 0 == rea_i ) {
+      break;
+    }
+    else {
+      dat_y += rea_i;
+      lef_i -= rea_i;
+    }
```
Nope. `if`-`else if`-`else` mirrors the structure in `c3_bile_put_raw()`, which is nice.

mcevoypeter commented on pkg/urbit/worker/mars.c:
```diff
@@ -696,272 +690,284 @@ _mars_timer_cb(uv_timer_t* tim_u)
+    mar_u->sav_u.tim_u.data = mar_u;
+  }
 
+  c3_path_free(lok_u);
```
Should the lock be released here?

mcevoypeter commented on pkg/urbit/worker/mars.c:
```diff
@@ -1152,6 +1158,7 @@ _mars_boot_make(u3_boot_opts* inp_u,
   }
 
   u3r_chubs(0, 2, met_u->who_d, who);
+  met_u->ver_w = 1; // elo_ver_w
```
Resolve.

mcevoypeter commented on pkg/urbit/worker/mars.c:
```diff
@@ -1272,73 +1279,113 @@ _mars_boot_make(u3_boot_opts* inp_u,
-  return c3y;
+end:
+  return suc_o;
+#undef create_dir
```
Remove.

mcevoypeter commented on pkg/urbit/worker/mars.c:
```diff
@@ -110,64 +119,57 @@ _mars_grab(u3_noun sac)
 
 /* _mars_gift(): enqueue response message.
 */
 static void
 _mars_gift(u3_mars* mar_u, u3_noun pro)
```
Decided against it. There are four call sites, which seems like enough to justify the branch.

mcevoypeter commented on pkg/urbit/vere/epoc.c:
```diff
@@ -0,0 +1,837 @@
+static inline c3_t
+_epoc_is_first(const c3_path* const pax_u)
+{
+  c3_c* sar_c = strstr(c3_path_str(pax_u), fir_nam_c);
+  return sar_c && 0 == strcmp(sar_c, fir_nam_c);
```
Decided against it.

mcevoypeter commented on pkg/urbit/vere/saga.c:
```diff
@@ -0,0 +1,749 @@
+  static const size_t siz_i = sizeof(((struct dirent*)NULL)->d_name);
+  const c3_c*         lef_c = *(const c3_c(*)[siz_i])lef_v;
+  const c3_c*         rih_c = *(const c3_c(*)[siz_i])rih_v;
+  const c3_i          dif_i = strlen(lef_c) - strlen(rih_c);
+  return 0 == dif_i ? strcmp(lef_c, rih_c) : dif_i;
```
After testing, it doesn't seem to be. Removed the check.

mcevoypeter commented on pkg/urbit/noun/events.c:
```diff
@@ -306,17 +318,14 @@ _ce_patch_read_control(u3_ce_patch* pat_u)
-  if ( (len_w != read(pat_u->ctl_i, pat_u->con_u, len_w)) ||
-        (len_w != sizeof(u3e_control) +
-                  (pat_u->con_u->pgs_w * sizeof(u3e_line))) )
+  if ( len_w != read(pat_u->ctl_i, pat_u->con_u, len_w)
+       || len_w != sizeof(u3e_control)
+                   + pat_u->con_u->pgs_w * sizeof(u3e_line) )
```
Yep, it is.

mcevoypeter commented on pkg/urbit/vere/saga.c:
```diff
@@ -0,0 +1,749 @@
+
+  if ( !u3_epoc_is_first(poc_u) ) {
+    cur_d = u3_epoc_first_evt(poc_u) <= u3A->eve_d
+              ? u3A->eve_d
```
Decided against it. It would complicate an already complicated interface.

mcevoypeter commented on pkg/urbit/vere/saga.c:
```diff
@@ -0,0 +1,749 @@
+  }
+
+  switch ( log_u->mod_e ) {
+    case u3_saga_sync:
+      if ( _epoc_is_full(poc_u) ) {
+        poc_u = _rollover(log_u);
+      }
+      log_u->eve_u.req_i = _request_len(poc_u, eve_u);
+      c3_lode* nod_u     = c3_list_peekf(eve_u);
+      log_u->act_t       = 1;
+      if ( !u3_epoc_commit(poc_u, nod_u, log_u->eve_u.req_i) ) {
+        goto fail;
+      }
+      _remove_committed_events(log_u);
+      log_u->act_t = 0;
+      goto succeed;
+    case u3_saga_async:
+      if ( !log_u->act_t ) {
+        if ( _epoc_is_full(poc_u) ) {
+          poc_u = _rollover(log_u);
+        }
+        log_u->eve_u.req_i = _request_len(poc_u, eve_u);
+        log_u->act_t       = 1;
+        uv_queue_work(log_u->asy_u.lup_u,
+                      &log_u->asy_u.req_u,
+                      _uv_commit_cb,
+                      _uv_commit_after_cb);
+      }
```
There are only four shared lines (excluding trivial lines), so it's not worth it.

matthew-levan commented:
@joemfb @mcevoypeter bumping this as it seems like the saga interface is a great starting point for seguro work. What's the status?

mcevoypeter commented:
@matthew-levan this is blocked by https://github.com/urbit/urbit/pull/5765, which is the backbone of the 1.9 runtime release. The 1.10 runtime release, which will include the mars-urth refactor and event log truncation, will go out shortly thereafter. @joemfb is focusing on the release process, so once 1.9 goes out the door, the priority will shift to reviewing and testing this PR and its intersection with the other work that's going into 1.10. Long story short, this should be reviewed, tested, and released in the very near future.

mcevoypeter commented:
Also, please let me know if you have feedback on the `saga` interface as it pertains to the Seguro work.

mcevoypeter commented:
> Address the nuances at the intersection of truncation and replaying the boot sequence. As things currently stand, the boot sequence must be replayed if the incremental snapshot is absent, which is clearly problematic if the first epoch, which contains the boot sequence has been dropped.

This issue mentioned in the PR description has been resolved. The `saga` module now exposes a `u3_saga_needs_bootstrap()` function, which determines whether the replay should attempt to replay the bootstrap sequence. The only remaining loose end is deciding on an optimal epoch size. The current epoch size is 100 events, which is much too small but handy for testing as rollover happens frequently.

matthew-levan commented:
Awesome thanks @mcevoypeter. Just wanted confirmation. I won't need this PR to be merged any time soon for the Seguro prototype to be completed, but it's good to have confirmation that I can use the `saga` interface in it without having to scrap it later.

zalberico commented:
@belisarius222  is going to take a look at the file to tell people not to edit files in the epoch directory.

@Fang- to partner with Joe and help with testing.

belisarius222 commented on pkg/urbit/tests/epoc_tests.c:
```diff
@@ -0,0 +1,148 @@
+  // the artifacts from a test run.
+#endif /* ifdef U3_EPOC_TEST */
+
+  fprintf(stderr, "epoc_tests: ok\r\n");
```
If these didn't run, because `U3_EPOC_TEST` wasn't defined, then it would be better to print something like `"epoc tests: skipped\r\n"`.

mcevoypeter replied:
> If I `#define U3_EPOC_TEST` in epoc.h, I get this when I run `make test`:
> 
> ```
> build/epoc_tests
> lmdb: failed to open environment at /tmp/event_log/0i138: No such file or directory
> epoc_tests: ok
> ```
> 
> Is that expected?

Yep, it is. The test is ensuring that attempting to open a non-existent epoch fails (see line `76` of `epoc_tests.c`).

belisarius222 commented:
If I `#define U3_EPOC_TEST` in epoc.h, I get this when I run `make test`:
```
build/epoc_tests
lmdb: failed to open environment at /tmp/event_log/0i138: No such file or directory
epoc_tests: ok
```

Is that expected?

mcevoypeter commented:
> If I `#define U3_EPOC_TEST` in epoc.h, I get this when I run `make test`:
> 
> ```
> build/epoc_tests
> lmdb: failed to open environment at /tmp/event_log/0i138: No such file or directory
> epoc_tests: ok
> ```
> 
> Is that expected?

Yep, it is. The test case is checking that attempting to open a non-existent epoch will fail (see line `76` of `epoc_tests.c`).

joemfb reviewed (changes requested):
My apologies for the many delays in getting this a full review. This PR is in very good shape. Outside of a couple minor things noted below, the behavior all looks to correctly implement its design.

I have some minor changes to that design (after reviewing it with the rest of the core team), and some missing features that will need to be implemented before release. The rest of my review focuses on correctness, crash-safety, clarity, boundaries between modules, and (of course) style, especially as pertains to existing code. I've tried to comment at least one instance of each type of feedback or request, but I've undoubtedly missed something.

Due to my delay and the intervening fixes and releases, this branch has conflicts with `next/vere`. Those will need to be resolved to finish a full review. That will be easiest if cosmetic changes to core headers are reverted, and some changes are simply reapplied after the merge (particularly those regarding the lockfile/log handling in `worker|daemon/main.c`).

##### needed features

- the version number of the binary needs to be recorded in each epoch

It's very important that we be able to identify binary upgrades purely from on-disk state, and for the creation of a new epoch
whenever the binary version changes. This lets us fix jet mismatches in place, and prepares us for future event log optimizations (commit-before-compute). The binary version string should be in a textfile, and reading it should trim trailing whitespace.

- a truncation command needs to be added

The people demand log truncation, and as soon as it's possible they'll do whatever is necessary to get it. The invariant is simple: all but the latest epoch can be safely removed. But they're all just sitting right next to each other, and `bash` snafus are to be expected ...

- the commit batch-size histogram should be restored

We need real statistics to prioritize event log implementations, and this is only a tiny bit of code.

##### design changes

- the binary files for pier metadata should be removed

Pier metadata is currently (on master) maintained in a second database, in the same environment as the event log. This was natural when there was only ever one event log database, but seems redundant and wasteful now that there will be several. I suggested the binary files, but I've never been particularly happy with them. The data is formally redundant given a log and/or snapshot, in every case except the lifecycle length.

  - identity: read by axis (specified by kelvin) from arvo's state, the sample of the arvo core
  - fake bit: ready via the namespace (from %jael, at a path specified by kelvin)
  - lifecycle length: specified by construction: the entire first epoch up to some max size (to handle migrated piers with arbitrarily-sized first epochs). `50` or `100` seem like reasonable upper bounds
  - epoch version: this should be text (`@ui`), with trailing whitespace trimmed before reading

- epoch creation should be ordered/optimized for crash recovery

Writing the initial snapshot is the most fragile part of creating an epoch, so it should be "guarded" by the version file (fully synced before the version is created). Under normal operation, events will be written immediately after creating an epoch, but is possible for a ship to remain idle and then crash, so an empty event log in the latest epoch isn't necessarily a sign of invalidity or corruption. And pier initialization (in `u3_saga_open()` and friends) will need to be modified to assume the invariants established by this ordering, and possible check/enforce them.

- epochs should be zero-indexed

Events cause state transitions, and the domain of their numbers is an enumeration of those states -- this is why there is no event `0`. Epochs are slices of these states: a snapshot establishing the initial state, and events causing subsequent transitions. Everyone asked has found this to be intuitively correct, and that's an important consideration; we'll be dealing with the implications of this design indefinitely.

- only the log of the latest epoch should be mapped in memory

Outside of initialization, older epochs are entirely inert. More speculatively, it might be worth explicitly making the files off all old epochs read-only on disk.

joemfb commented on pkg/urbit/worker/mars.c:
```diff
@@ -1272,77 +1278,110 @@ _mars_boot_make(u3_boot_opts* inp_u,
+//! @n (2) Commit boot events to the event log.
+//! @n (3) Serialize boot event noun into buffer.
+//! @n (4) First four bytes are for mug, and the mug for boot events is 0.
+//! @n (5) Create single event out of boot events and attempt to boot.
```
For posterity: https://github.com/urbit/urbit/pull/5940#discussion_r948162614

Fang- commented on pkg/urbit/vere/saga.c:
```diff
@@ -0,0 +1,730 @@
+  return epo_min_d == u3_epoc_first_commit(poc_u) && 1 == len_i; // (1)
+#else
+  return 1; // (2)
+# endif /* ifndef U3_REPLAY_FULL */
```
From looking at `_read_epoc_dirs` I imagined it was, and I _think_ all epoch list loading happens through the one callsite for that?

mcevoypeter replied:
@Fang- is right: the epoch list is sorted (and yes, all epoch loading is done through `_read_epoc_dirs()`). This condition checks "is our oldest epoch the first epoch *and* is it the only epoch we have". We only need to bootstrap if both conditions are true. If the former is false, then we can boot from our oldest epoch's snapshot (in the worst case) because it's not the first epoch, which doesn't have a snapshot. If the latter is false, then we can boot from some more recent epoch. This of course is only relevant in the context of truncation, which isn't implemented via the binary yet but does work in the obvious way of `rm -rf <pier>/.urb/log/0i1`.

Fang- commented:
This week I've been working on adding logic for "pier validation" on top of this PR. I'll offer some brief, surface-level thoughts in addition to @joemfb's deeper review (which I fully +1).

I was a bit surprised to find some purely stylistic changes when first looking at the diff here, I'll second that this makes review harder and indeed makes conflicts more likely. I also agree that new style goes onto new code first, to give it time to simmer and be adopted (or rejected) by other people.

As a small-time vere contributor, and that being the extent of my real-world C experience, the new style is certainly showing me things I wasn't aware you could do. In some places this is cute, in other places this is confusing.  
Jumping in, it certainly felt like there was more indirection & abstraction than I was used to from vere. This goes for both reading the code (not least of which `@n` lol), and writing it. I had written some instances of `poc_u->las_d` before realizing I _should_ be using `u3_epoc_last_commit` instead.

The sheer straightforwardness of old/current vere-style C appeals to me, and it feels like we're leaving some of that behind. Perhaps that's the right call to make, as the codebase and the team working on it evolves. And note, I'm always slow & reluctant about change, so make of this what you will.

All that said, this is a lot of good work. I'm really excited about this finally becoming real, and perhaps even more excited to just save money on disk space. (^:

mcevoypeter commented on pkg/urbit/worker/mars.c:
```diff
@@ -696,272 +700,273 @@ _mars_timer_cb(uv_timer_t* tim_u)
-    return 0;
-  }
+  c3_path* const lok_u = c3_path_fp(mar_u->dir_u);
+  u3_lock_acquire(lok_u);
```
Copying `mar_u->dir_u` simplified error handling by removing the need to `c3_path_pop(mar_u->dir_u)` twie in the case where opening the event log failed. To simplify, I've replaced `lok_u` with `dir_u`, which represents the path to the event log, which should also address https://github.com/urbit/urbit/pull/5701#discussion_r957693722.

mcevoypeter commented on pkg/urbit/worker/mars.c:
```diff
@@ -696,272 +700,273 @@ _mars_timer_cb(uv_timer_t* tim_u)
-    u3e_save();
-  }
+    c3_path_push(mar_u->dir_u, ".urb");
+    c3_path_push(mar_u->dir_u, "log");
```
Agreed with respect to mutating `mar_u->dir_u`. I've addressed this as described in https://github.com/urbit/urbit/pull/5701#discussion_r959934204.

mcevoypeter commented on pkg/urbit/vere/saga.c:
```diff
@@ -0,0 +1,730 @@
+  }
+
+  { // (2)
+    try_list(log_u->epo_u.lis_u = c3_list_init(), goto free_event_log);
```
Omitting the `try_` would leak the implementation details of `c3_list_init()`. Creating a list is an operation that can logically fail, and it's in our best interest to not make any assumptions about how that failure occurs when using the `c3_list` interface.

joemfb replied:
This feels like null-hypothesis juggling. `c3_list` is presumably part of the `c3` "package", and allocations in `c3` do not locally fail. Even if that counts as an "implementation detail", the question is: what should the `c3_list` API be? Specifying that it can fail when it actually cannot requires callers to include non-operational boilerplate, which further complicates already dense and subtle code.

mcevoypeter commented on pkg/urbit/worker/mars.c:
```diff
@@ -1272,77 +1278,110 @@ _mars_boot_make(u3_boot_opts* inp_u,
-    return c3n;  //  XX cleanup
+  u3_noun evt = u3_nul;
+  if ( !u3_saga_replay(log_u, 0, met_u.lif_w, _saga_boot_cb, &evt) ) { // (5)
+    goto free_event_log;
```
I'm not sure if I follow. `u3_saga_replay()` iterates over a specified subset of events in the event log, invoking the provided callback with each event. By replaying the first `lif_w` events, we're able to construct the boot sequence and then execute it, which is a mechanism that feels more elegant than a raw iterator (like `epoc` has). Are you saying that each boot event has to be individually computed for the `boot` IPC command to work correctly? If the replay of the boot sequence fails here, `u3_mars_boot()` will return `c3n`, indicating that boot failed.

mcevoypeter replied:
This was a misunderstanding on my part. The life cycle length refers to the length of the initial subset of events in the boot sequence that are required to minimally bootstrap an Arvo kernel. Replay at boot should not just replay these events but the entire boot sequence.

mcevoypeter commented on pkg/urbit/include/vere/meta.h:
```diff
@@ -0,0 +1,18 @@
+  c3_d who_d[2]; //!< identity
+  c3_o fak_o;    //!< fake bit
+  c3_w lif_w;    //!< lifecycle length
+  c3_w ver_w;    //!< event log version
```
This is the event log version number, not the epoch version number. Each epoch also has a version number, although there is no field in the `u3_epoc` struct at the moment. See `epo_ver_w` in `epoc.c`. I assumed that it would be useful to have a version number on the event log too in case we, say, make interface changes to `saga`.

joemfb replied:
Where is this number sourced from. Is there a version file in `log/` in addition to each `log/$epoch`?

joemfb replied:
Bump.

mcevoypeter replied:
Sorry about the delay here. This thread got lost in the shuffle. This use of `ver_w` was a mistake on my part. I intended it to represent the event log version number but failed to persist it, so it was essentially useless. Because each epoch is versioned, I think we can get away without a version number on the event log itself, so I've removed it. What do you think?

mcevoypeter commented on pkg/urbit/worker/mars.c:
```diff
@@ -696,272 +700,273 @@ _mars_timer_cb(uv_timer_t* tim_u)
-    //
+  // XX refactor |mass
+  // XX get batch size from args
+  static const c3_d bat_d = 1;
```
Bumped to 500.

mcevoypeter commented on pkg/urbit/worker/mars.c:
```diff
@@ -696,272 +700,273 @@ _mars_timer_cb(uv_timer_t* tim_u)
 
-  u3l_log("---------------- playback complete ----------------\r\n");
+  if ( las_d == cur_d ) {
+    fprintf(stderr, "---------------- playback complete ----------------\r\n");
```
Noted! Thanks for pointing this out.

joemfb commented on pkg/urbit/vere/epoc.c:
```diff
@@ -0,0 +1,903 @@
+    do {                                                                       \
+      MDB_val key_u = {                                                        \
+        .mv_data = (void*)key_string,                                          \
+        .mv_size = strlen(key_string) + 1,                                     \
```
This is wrong, the existing metadata keys are byte arrays without the null terminator.

mcevoypeter replied:
Thanks for pointing this out. I've corrected the issue.

joemfb commented on pkg/urbit/vere/epoc.c:
```diff
@@ -0,0 +1,903 @@
+
+  goto succeed;
+
+abort_txn:
+  mdb_txn_abort(txn_u);
+rename_lock_mdb:
+  c3_assert(_move_file(poc_u->pax_u, src_u, "lock.mdb"));
+rename_data_mdb:
+  c3_assert(_move_file(poc_u->pax_u, src_u, "data.mdb"));
+free_epoc:
+  u3_epoc_close(poc_u);
+  c3_free(poc_u);
+fail:
+  return NULL;
```
I'm concerned about this approach to failed migrations. Moving files first -- and moving them back if anything is missing/incomplete or if anything goes wrong -- is more fragile than collecting and validating all data before moving the files. I/O can fail, and if it is at all possible, we want to leave the pier undisturbed when we can't migrate.

mcevoypeter replied:
This has also been corrected.

joemfb reviewed (changes requested):
The migration is currently broken due to mishandled metadata keys. That stuff will need to change anyway, since we're moving away from the pier metadata approach (and will be extracting data from arvo directly).

I have some more general comments about the migration design that I forgot to make in my previous review:

mcevoypeter commented on pkg/urbit/noun/manage.c:
```diff
@@ -1765,17 +1768,15 @@ u3m_stop()
 */
 c3_d
-u3m_boot(c3_c* dir_c)
+u3m_boot(const c3_c* dir_c, c3_o (*sap_f)(const c3_c* dir_c))
```
Good point. I've reverted the `u3m_boot()` interface back to its previous state and added a `_boot_from_epoc_snapshot()` in `saga.c`.

mcevoypeter commented on pkg/urbit/noun/manage.c:
```diff
@@ -1737,17 +1738,19 @@ u3m_init(void)
     }
 
-    u3l_log("loom: mapped %dMB\r\n", len_w >> 20);
+    fprintf(stderr, "loom: mapped %dMB\r\n", len_w >> 20);
```
Addressed by the change mentioned in https://github.com/urbit/urbit/pull/5701#discussion_r961027361.

mcevoypeter commented on pkg/urbit/noun/events.c:
```diff
@@ -689,91 +797,123 @@ _ce_patch_apply(u3_ce_patch* pat_u)
-    }
+    c3_assert(-1 != read(pat_u->mem_i, mem_w, pag_siz_i));
+    c3_assert(-1 != lseek(fid_i, off_w * pag_siz_i, SEEK_SET));
+    c3_assert(-1 != write(fid_i, mem_w, pag_siz_i));
```
Good point. Reverted back to the old code.

mcevoypeter commented on pkg/urbit/worker/mars.c:
```diff
@@ -696,272 +700,273 @@ _mars_timer_cb(uv_timer_t* tim_u)
+    try_saga(mar_u->log_u = u3_saga_open(mar_u->dir_u, &mar_u->met_u),
+             goto release_lock,
+             "failed to open event log at %s",
+             c3_path_str(mar_u->dir_u));
```
Perhaps unsurprisingly, I disagree. Each module that has a `try_<module_name>` macro (i.e. `saga`, `epoc`, etc) uses consistent return types in its interface. That is, with few exceptions, zero conveys failure and non-zero conveys success. As a result, the majority of use cases of these interfaces obeys a well-defined pattern. Furthermore, each of the `try_*` macros work in exactly the same way. So the use of the macros reduces the cognitive energy required to read the code because there are fewer moving pieces around a call into one of these interfaces. Your example is harder to read because there's more slack i.e. more spaces in which arbitrary code can be inserted, which makes it harder to understand. Using `try_saga`, we have
```c
try_saga(mar_u->log_u = u3_saga_open(mar_u->dir_u, &mar_u->met_u), goto release_lock, /* snip */);
```
which unambiguously reads as "attempt to open the event log and assign it to `mar_u->log_u`, and if that fails, `goto release_lock`. In other words, the `try_*` macros reduce the number of branches that someone reading the code needs to hold in their head at any one time because many of the branches are identical, which allows them to be collapsed into a single pattern by the relevant `try_*` macro. And in the case where more complex logic is required, an `if` statement as you describe will always work *and*, by breaking the `try_*` pattern, will obviously convey that something different is going on.

mcevoypeter commented on pkg/urbit/vere/saga.c:
````diff
@@ -0,0 +1,730 @@
+//! <epoch_M>/
+//! ```
+//! Note that the epoch directory names are ommitted because they are an
+//! implementation detail of the epoc module.
````
Sure, raw epoch numbers are an interface, but they're not a particularly flexible interface. `u3_epoc_first_commit()`, `u3_epoc_last_commit`(), and `u3_epoc_first_evt()` are more flexible (and more obviously named) interfaces. The only guarantee that the `epoc` module needs to make with respect to naming is that epoch names obey the same total ordering that their underlying events obey.

mcevoypeter commented on pkg/urbit/vere/epoc.c:
```diff
@@ -0,0 +1,915 @@
+
+  if ( epo_min_d != fir_d ) { // (4)
+#ifndef U3_EPOC_TEST
+    u3e_save();
```
I don't necessarily disagree, but why do you think so?

mcevoypeter commented on pkg/urbit/vere/saga.c:
```diff
@@ -0,0 +1,730 @@
+                      _uv_commit_after_cb);
+      }
+      goto succeed;
+  }
```
Good point. I've factored `u3_saga_commit()` into `u3_saga_commit_sync()` and `u3_saga_commit_async()`.

mcevoypeter commented on pkg/urbit/vere/saga.c:
```diff
@@ -0,0 +1,730 @@
+u3_saga_needs_rollover(const u3_saga* const log_u)
+{
+  return u3_epoc_len(log_u->epo_u.cur_u) >= epo_len_i;
+}
```
I'm not sure I agree. The size of the event log can (and should) be determined within the `saga` module, and any runtime config can be threaded through if necessary.

mcevoypeter commented on pkg/urbit/vere/saga.c:
```diff
@@ -0,0 +1,730 @@
+               c3_d           cur_d,
+               c3_d           las_d,
+               u3_saga_play   pla_f,
+               void*          ptr_v)
```
The abstraction it provides is not "replay", but "restore the ship to latest snapshot and then invoke a user-provided function for each event that occurred after that snapshot," which I'd argue is a valuable abstraction. Again, the emphasis here is flexibility. It shares the vast majority of the replay functionality--the snapshot restoration and subsequent event iteration--without sacrificing the ability to take arbitrary action on each replayed event. So yes, replaying the boot sequence and replaying non-boot-sequence events require different callbacks, but I don't see the problem with that. This mechanism will also let us more easily analyze the events in the most recent epoch, should we need to do so.

mcevoypeter commented on pkg/urbit/worker/mars.c:
```diff
@@ -582,30 +584,30 @@ _mars_flush(u3_mars* mar_u)
+          && !u3_saga_rollover(mar_u->log_u) )
+      {
+        fprintf(stderr, "mars: failed to rollover to new epoch\r\n");
+      }
```
Again, this movement of event log specific logic out of `saga` and into `mars` worries me. If we're not careful, instead of two modules with clear responsibilities, we'll end up with a single, much bigger module that's much more difficult to understand. `mars` is the state machine, and while the event log is a vital component of that state machine, it's not small enough to be subsumed within the state machine, especially if we hope to be able to swap out event log implementations for the sake of hosting.

mcevoypeter commented:
Thank you @joemfb, @Fang-, and @belisarius222 for the feedback. I've addressed the top-level needed features and design changes, specifically:
- The binary version number is recorded in each epoch.
- A truncation command--`urbit chop`--has been added.
- The commit batch-size histogram has been implemented.
- Binary metadata files have been removed.
- Epochs are now zero-indexed.
- Only the two most recent epochs (two is needed to account for the edge case in which the most recent epoch is empty) are mapped into memory.

I'll be moving on to the style feedback on Monday, which I expect to take much less time than the functionality changes listed above.

mcevoypeter commented on pkg/urbit/noun/events.c:
```diff
@@ -782,96 +922,49 @@ _ce_image_fine(u3e_image* img_u,
-  if (  (-1 == lseek(fom_u->fid_i, 0, SEEK_SET))
-     || (-1 == lseek(tou_u->fid_i, 0, SEEK_SET)) )
+  if ( (-1 == lseek(fom_u->fid_i, 0, SEEK_SET))
+       || (-1 == lseek(tou_u->fid_i, 0, SEEK_SET)) )
```
> Incrementally restyling existing code while also changing functionality makes it harder to diff, review, or trace the history.

I agree, and at the same time, improving code hygiene is an incremental process. Our code style is inconsistent, particularly around whitespace, which is why I proposed the introduction of a linter in https://github.com/urbit/urbit/pull/5481. Because I met with resistance to linting, I've slipped into the habit of applying a more consistent style in whatever file I'm working in. I acknowledge that style changes intermingled with functionality is a pain, but so is a lack of a consistent coding style.

I've backed out the intrusive style changes, particularly in the `events` module. That said, you'll notice that I didn't revert the documentation changes in the `events` or `manage` modules because those changes provide more explicit explanation of what the code is doing and/or presents a more organized method of documenting. I also have refrained from backing out the replacement of bitwise page arithmetic with general arithmetic in `events` because I've now made numerous substantial changes to the `events` module, each of which has forced me to spend more time thinking about the math than I'd like to, which I attribute largely to the use of bitwise arithmetic. Furthermore, any compiler worth its salt will replace multiplication/division with a bit shift whenever possible, so using general arithmetic gets us the best of both worlds: the improved readability of general arithmetic and improved performance of bitwise arithmetic.

mcevoypeter commented on pkg/urbit/noun/events.c:
The only change here was from use of `//!` to `///`. `///` behaves better with auto-indent in `neovim` (and presumably other editors).

joemfb commented on pkg/urbit/vere/saga.c:
```diff
@@ -0,0 +1,730 @@
+    u3_epoc_info(poc_u);
+    nod_u = c3_lode_next(nod_u);
+  }
+}
```
I see that the histogram is now being maintained -- the stored values should also be printed here.

mcevoypeter replied:
Just to be explicit, I added both the histogram and the logging of its values in `u3_saga_info()`.

joemfb commented on pkg/urbit/include/vere/epoc.h:
```diff
@@ -0,0 +1,272 @@
+const c3_c*
+u3_epoc_path_str(const u3_epoc* const poc_u);
+
+/// Get the ID of the first committed event in an epoch.
+///
+/// @param[in] poc_u  Epoch handle. Must have at least one committed event and
+///                   must not be NULL.
+///
+/// @return  ID of epoch's first committed event.
+c3_d
+u3_epoc_first_commit(const u3_epoc* const poc_u);
+
+/// Get the ID of the last committed event in an epoch.
+///
+/// @param[in] poc_u  Epoch handle. Must have at least one committed event and
+///                   must not be NULL.
+///
+/// @return  ID of epoch's last committed event.
+c3_d
+u3_epoc_last_commit(const u3_epoc* const poc_u);
+
+/// Get the ID of the first event represented by the epoch, which is equivalent
+/// to the ID of the event that corresponds to the epoch's snapshot.
+///
+/// @param[in] poc_u  Epoch handle. Must not be NULL.
+///
+/// @return  ID of epoch's first event.
+c3_d
+u3_epoc_first_evt(const u3_epoc* const poc_u);
```
This API is a little confusing. "first_evt" is actually just the epoch number, and it refers to the initial state of the epoch (stored in the snapshot, caused by the last event of the preceding epoch). First/last "commit" is actually the first and last event in the epoch.

joemfb reviewed (changes requested):
Some further tweaks and follow-ups:

mcevoypeter commented on pkg/urbit/worker/mars.c:
```diff
@@ -995,87 +964,122 @@ u3_mars_init(c3_c*    dir_c,
+    // Scry sample for reading fake bit is [[~ ~] [%.n [%once %j %fake ~]]]
+    u3_cell sam = u3nc(u3nc(u3_nul, u3_nul),
+                       u3nc(c3n, u3nq(c3__once, 'j', c3__fake, u3_nul)));
+    u3_noun res = u3v_soft_peek(1000 /* ms */, sam);
```
Disabled the timeout.

mcevoypeter commented on pkg/urbit/worker/mars.c:
```diff
@@ -624,13 +625,21 @@ _mars_flush(u3_mars* mar_u)
-      //  XX dispose [mar_u], exit cb ?
-      //
+      c3_free(mar_u->dir_u);
+      for ( size_t idx_i = 0; idx_i < c3_list_len(mar_u->gif_u); idx_i++ ) {
```
Good point.

mcevoypeter commented:
In response to https://github.com/urbit/urbit/pull/5701#discussion_r975622819:

Allocations in `c3` do not locally fail, but if that were to change for whatever reason and we were to make `c3_list_init()` unconditionally succeed, then we'd have a much bigger problem on our hands: changing all call sites of `c3_list_init()` to handle the newly-possible failure, which would be much more painful than the inclusion of a minimal amount of boilerplate at the current call sites. Memory allocation is an operation that can logically fail, and in an interface as general as `c3_list`, we should preserve the ability to fail because it gives us more flexibility.

mcevoypeter commented on pkg/urbit/worker/mars.c:
```diff
@@ -591,30 +592,30 @@ _mars_flush(u3_mars* mar_u)
-    if ( !mar_u->gif_u.ext_u ) {
-      mar_u->gif_u.ent_u = 0;
+      c3_free(c3_list_popf(mar_u->gif_u));
+      gif_u = c3_lode_data(c3_list_peekf(mar_u->gif_u));
```
This actually doesn't leak `gif_u->hun_y`. `gif_u->hun_y` is passed to `u3_newt_send()`, which assigns it to the `buf_y` field of `n_req* req_u`. In `_newt_write_cb`, `req_u->buf_y` is freed once the write completes.

mcevoypeter replied:
So freeing `gif_u->hun_y` here results in a double free, leading the runtime to crash when attempting to boot a new fake ship (at least when I tried to do so on my machine).

joemfb commented on pkg/urbit/worker/mars.c:
```diff
@@ -110,64 +118,57 @@ _mars_grab(u3_noun sac)
+                         dat_y,
+                         len_i);
   }
+
```
This should go back to commit first, then queue effects. If the event log is idle, that async commit can begin earlier that way.

mcevoypeter commented on pkg/urbit/worker/mars.c:
```diff
@@ -110,64 +118,57 @@ _mars_grab(u3_noun sac)
-    u3z(pro);
+    c3_l   mug_l = mar_u->mug_l;
+    size_t len_i = sizeof(mug_l) + len_w;
+    c3_y   dat_y[len_i];
```
To address this, I added the notion of ownership semantics to `c3_list`, where a list node can either copy the caller-provided data when a new node is pushed or simply take over ownership of that data.

mcevoypeter commented on pkg/urbit/worker/mars.c:
```diff
@@ -995,87 +964,122 @@ u3_mars_init(c3_c*    dir_c,
+      goto free_event_log;
+    }
+    u3r_bytes(0, u3r_met(3, who), (c3_y*)&mar_u->met_u.who_d, who);
+  }
```
Thanks for pointing this out. I've added the version switch.

mcevoypeter commented on pkg/urbit/worker/mars.c:
```diff
@@ -110,64 +119,57 @@ _mars_grab(u3_noun sac)
-*/
+ */
+
+//! @n (1) TRANSFER.
```
Good to know. The `TRANSFER` comment has been removed.

mcevoypeter commented:
@joemfb, I think that I've addressed all of your feedback, so it's ready for another look from you. Below are the two main comments for which I'd appreciate further clarification:

- https://github.com/urbit/urbit/pull/5701#discussion_r958801588
- https://github.com/urbit/urbit/pull/5701#discussion_r957846678

xiphiness commented:
can i build this and prune my event log without getting fucked by incompatibility if something changes i really really need this can devs pls do something im desperate if i increase my storage costs i won't be able to feed my family


(greatly appreciate all the hard work that went into this)

zalberico commented:
Can you hold on a little longer? Shipping this feature is our primary focus on the runtime currently and we can push more updates here soon.

xiphiness commented:
I mean, theoretically, yes. but I'm mainly curious about what state its in from the ppl working on it, because if it's in a mature enough state that it will be forwards compatible w new release I would just build it myself

xiphiness commented:
to put things in context: I had so little space left on my device all 10 ships I host crashed and I don't even have enough space for bash to autocomplete. one guy is willing to self host and there's a backup I can prune, giving me 9GB, but I'm in pretty dire straits

zalberico commented:
Unfortunately, the work this PR was built on top of (the mars/"urth" refactor) had a show stopping bug that required the underlying work in it to get broken down and shipped in pieces (I can't find any details about this bug anywhere in github - if there's a link please add it here).

As a result, this PR cannot ship as is and would need to be reimplemented on top of the current vere.

The underlying work this PR is built on that can be safely pulled out and shipped independently from the mars/urth refactor is being pulled out and shipped separately (memory corruption fixes, demand paging, other fixes in the 1.14rc).

Joe is working on a minimal command line version of event log truncation in the mean time which should help us be unblocked and we'll see if that's able to ship before he's out. 

We'll take a look at this live implementation work in the new runtime repo once we have that doing releases.
